src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could close wireguard container: %m");
+
+          r = sd_netlink_send(netdev->manager->genl, message, &serial);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could not set wireguard device: %m");
+
+        } while (peer || mask_start);
+
+        return 0;
+}
+
+static WireguardEndpoint* endpoint_free(WireguardEndpoint *e) {
+        assert(e);
+        netdev_unref(e->netdev);
+        free(e->host);
+        free(e->port);
+        free(e);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(WireguardEndpoint*, endpoint_free);
+
+static int on_resolve_retry(sd_event_source *s, usec_t usec, void *userdata) {
+        NetDev *netdev = userdata;
+        Wireguard *w;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        w->resolve_retry_event_source = sd_event_source_unref(w->resolve_retry_event_source);
+
+        w->unresolved_endpoints = w->failed_endpoints;
+        w->failed_endpoints = NULL;
+
+        resolve_endpoints(netdev);
+
+        return 0;
+}
+
+static int wireguard_resolve_handler(sd_resolve_query *q,
+                                     int ret,
+                                     const struct addrinfo *ai,
+                                     void *userdata) {
+        NetDev *netdev;
+        Wireguard *w;
+        _cleanup_(endpoint_freep) WireguardEndpoint *e;
+        int r;
+
+        assert(userdata);
+        e = userdata;
+        netdev = e->netdev;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_REMOVE(endpoints, w->unresolved_endpoints, e);
+
+        if (ret != 0) {
+                log_netdev_error(netdev, "Failed to resolve host '%s:%s': %s", e->host, e->port, gai_strerror(ret));
+                LIST_PREPEND(endpoints, w->failed_endpoints, e);
+                e = NULL;
+        } else if ((ai->ai_family == AF_INET && ai->ai_addrlen == sizeof(struct sockaddr_in)) ||
+                        (ai->ai_family == AF_INET6 && ai->ai_addrlen == sizeof(struct sockaddr_in6)))
+                memcpy(&e->peer->endpoint, ai->ai_addr, ai->ai_addrlen);
+        else
+                log_netdev_error(netdev, "Neither IPv4 nor IPv6 address found for peer endpoint: %s:%s", e->host, e->port);
+
+        if (w->unresolved_endpoints) {
+                resolve_endpoints(netdev);
+                return 0;
+        }
+
+        set_wireguard_interface(netdev);
+        if (w->failed_endpoints) {
+                w->retries++;
+                r = sd_event_add_time(netdev->manager->event,
+                                      &w->resolve_retry_event_source,
+                                      CLOCK_MONOTONIC,
+                                      /* cap at ~25s */
+                                      now(CLOCK_MONOTONIC) + (2 << MAX(w->retries, (unsigned)7)) * 100 * USEC_PER_MSEC,
```

Comment:
I don't know systemd's policies on this kind of thing. Is 25s reasonable? In `wg(8)` I try for 4.29 minutes. I start by waiting 1 second, and then keep multiplying by 1.5 until I'm waiting for 90 seconds between tries, and then I give up. Not sure if this is even a reasonable approach itself, but I'd be interested in input here.

Reply:
(unsigned) 7 → 7U...

Also, could this calculation please be moved to its own little functoin with plenty comments what is going in here?